ClassFieldTheory/Cohomology/IsFilterComplete.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+import Mathlib.Algebra.BigOperators.Intervals
+import Mathlib.Algebra.Group.Subgroup.Pointwise
+import Mathlib.Algebra.Group.Submonoid.BigOperators
+import Mathlib.GroupTheory.QuotientGroup.Defs
+
+/-! # Completeness with respect to a filtration -/
+
+-- to replace `IsHausdorff`
+/-- `M` is Hausdorff with respect to the filtration `M_i ≤ M` if `⋂ M_i = 0`. -/
+class IsFilterHasudorff {M ι σ : Type*} [AddCommGroup M]
+    [LE ι] [SetLike σ M] [AddSubgroupClass σ M] (F : ι → σ) : Prop where
+  haus' : ∀ x : M, (∀ i, x ∈ F i) → x = 0
+
+-- to replace `IsPrecomplete`
+/-- `M` is precomplete with respect to the filtration `M_i ≤ M` if any compatible
+sequence `I → M` has a limit in `M`. -/
+class IsFilterPrecomplete {M ι σ : Type*} [AddCommGroup M]
+    [LE ι] [SetLike σ M] [AddSubgroupClass σ M] (F : ι → σ) : Prop where
+  prec' : ∀ x : (ι → M), (∀ i j, i ≤ j → x i - x j ∈ F i) → ∃ L : M, ∀ i, x i - L ∈ F i
+
+-- to replace `IsAdicComplete`
+/-- `M` is complete with respect to the filtration `M_i ≤ M` if `⋂ M_i = 0` and any compatible
+sequence `I → M` has a limit in `M`. -/
+@[mk_iff] class IsFilterComplete {M ι σ : Type*} [AddCommGroup M]
+    [LE ι] [SetLike σ M] [AddSubgroupClass σ M] (F : ι → σ) : Prop
+    extends IsFilterHasudorff F, IsFilterPrecomplete F
+
+section
+variable {M ι σ : Type*} [AddCommGroup M] [LE ι] [SetLike σ M] [AddSubgroupClass σ M] (F : ι → σ)
+
+def Completion.set : Set ((i : ι) → M ⧸ AddSubgroup.ofClass (F i)) :=
+  {f | ∀ ⦃i j⦄, i ≤ j →
+    QuotientAddGroup.map _ (AddSubgroup.ofClass (F i) ⊔ .ofClass (F j)) (.id M) le_sup_left (f i) =
+    QuotientAddGroup.map _ _ (.id M) le_sup_right (f j)}
+
+/-- Completion that works for different classes such as `Module` or `AddCommGroup`. -/
+def Completion : Type _ := Completion.set F
+
+namespace Completion
+
+def addSubgroup : AddSubgroup ((i : ι) → M ⧸ AddSubgroup.ofClass (F i)) where
+  carrier := set F
+  zero_mem' _ _ _ := by simp
+  add_mem' ha hb _ _ h := by simpa using congr($(ha h) + $(hb h))
+  neg_mem' ha _ _ h := by simpa using congr(-$(ha h))
+
+instance : AddCommGroup (Completion F) := addSubgroup F |>.toAddCommGroup
+
+end Completion
+
+def toCompletion : M →+ Completion F where
+  toFun x := ⟨fun _ ↦ x, fun _ _ _ ↦ rfl⟩
+  map_zero' := rfl
+  map_add' _ _ := rfl
+
+theorem toCompletion_injective_iff_isFilterHausdorff :
+    Function.Injective (toCompletion F) ↔ IsFilterHasudorff F :=
+  sorry -- may need more assumptions?
+
+theorem toCompletion_surjective_iff_isFilterPrecomplete :
+    Function.Surjective (toCompletion F) ↔ IsFilterPrecomplete F :=
+  sorry -- may need more assumptions?
```

Comment:
This one I think is false, at least I got stuck on the proof:
```
theorem toCompletion_surjective_iff_isFilterPrecomplete :
    Function.Surjective (toCompletion F) ↔ IsFilterPrecomplete F := by
  constructor
  · intro h
    constructor
    intro f hf
    specialize h ⟨fun j ↦ f j, ?_⟩
    · intro i j h
      specialize hf i j h
      simp
      rw [← QuotientAddGroup.mk_eq_mk_iff] at hf
      /-
      h : i ≤ j
      hf : ↑(f i) = ↑(f j)
      ⊢ ↑(f i) = ↑(f j)

      but goal is equality in M / (F i + F j) and hypothesis is equality in M / F i
      which is weaker
      -/
      sorry
    · sorry
  · sorry
```
We know i <= j but can't deduce F j <= F i which is what I seemed to need. I'm reluctant to merge with a false sorry.

Reply:
done